### ISSUE TITLE
P1-18 degenerate program mismatches the f64 reference: relative L2 error 1.08 (limit 0.0039)

### ISSUE THREAD

**Issue by gpillon** [gpillon/ignis#74]
## Context

`crates/core/tests/step_degenerate_gpu.rs` (P1-18, #54) checks the degenerate
step ABI — embedding -> final RMSNorm -> W8G32 output head -> argmax, every
decoder layer skipped — against an independent from-scratch Rust f64 decode
of the artifact's container bytes.

This test has never actually executed before tonight: it was missing
`#[ignore]`, so `scripts/gpu-profile.ps1`'s exact invocation
(`cargo test --features cuda -- --ignored`) silently filtered it out every
run (#73, now fixed). Running it for the first time on a free RTX 5090
(2026-09-07):

```
thread 'degenerate_program_matches_f64_reference_for_four_tokens' panicked at crates\core\tests\step_degenerate_gpu.rs:193:5:
token 0: relative L2 error 1.0759044333675807 exceeds 0.00390625
```

The tolerance (`assert_matches_bf16_tolerance`) is `BF16_UNIT_ROUNDOFF = 1/256
= 0.00390625` on the relative L2 error over the full 248320-entry logits
vector. An error of ~1.08 is not a rounding-level mismatch — the GPU-computed
logits for token 0 are essentially uncorrelated with the independent f64
reference.

## Scope

The other two previously-invisible tests fixed by #73 both pass on the same
run: `gdn_layer_matches_f64_reference` (P1-22) and
`real_nvfp4full_model_load_binds_every_text_scope_object` (P1-17). This
failure is isolated to the degenerate program's embedding -> norm -> output
head -> argmax path (P1-18) — the narrowest layer-skipping slice of the
forward, so the fault domain should be small: embedding gather, final
RMSNorm, or the W8G32 output head GEMV/dequant, or a mismatch between the
topology descriptor and what the test's from-scratch f64 decoder assumes
about the container layout.

## Blocks

G1 (#36 / #62): this is one of the acceptance-criteria checks named directly
by the gate ("the P1-21/22 layer checks... are green" implies the P1-18
program-slice check must hold too, since P1-21/22 build on it). Only
confirmed run to date is red.

**Comment by gpillon** [comment]
Root cause: the test's own from-scratch f64 oracle, not the kernel. `text/final_norm` is stored in unit-offset (`1+w`) form, same as every other hidden-width RMSNorm gain (`crates/artifact/tests/rmsnorm_weight_convention.rs`, #67), and the device program's `unit_offset=true` call (kernel/src/step.cu) is already correct — the oracle in `step_degenerate_gpu.rs` just forgot the offset, producing the observed relative L2 1.08. Fixed in 4834eea, verified green (`degenerate_program_matches_f64_reference_for_four_tokens ... ok`, 54.11s) on a free RTX 5090, merged to main in 767bee7.
